src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor/RazorUpdaterStep.cs

```diff
@@ -0,0 +1,148 @@
+        }
+
+        /// <summary>
+        /// Determines whether the RazorUpdaterStep applies to a given context. This doesn't
+        /// consider whether there's actually any work to be done by Razor updaters. Rather, it
+        /// determines whether it makes sense to even initialize the Razor updaters. So, it
+        /// will return false if there are no cshtml files in the current project or if there
+        /// are no Razor updaters available.
+        /// </summary>
+        /// <param name="context">The context to evaluate.</param>
+        /// <param name="token">A token that can be used to cancel execution.</param>
+        /// <returns>True if the Razor updater step might apply, false otherwise.</returns>
+        protected override Task<bool> IsApplicableImplAsync(IUpgradeContext context, CancellationToken token) =>
+            Task.FromResult(context?.CurrentProject is not null
+            && SubSteps.Any()
+            && GetRazorFileSystem(context.CurrentProject.Required()).EnumerateItems("/").Any());
+
+        protected override async Task<UpgradeStepInitializeResult> InitializeImplAsync(IUpgradeContext context, CancellationToken token)
+        {
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            _razorEngine = RazorProjectEngine.Create(RazorConfiguration.Default, GetRazorFileSystem(context.CurrentProject.Required()), builder =>
+            {
+                // Add useful "MVC View-isms" like _ViewImports support, page models, etc.
+                // https://github.com/dotnet/aspnetcore/blob/main/src/Razor/Microsoft.AspNetCore.Mvc.Razor.Extensions/src/RazorExtensions.cs
+                RazorExtensions.Register(builder);
+            });
+
+            ProcessRazorDocuments();
+
+            foreach (var step in SubSteps)
+            {
+                await step.InitializeAsync(context, token).ConfigureAwait(false);
+            }
+
+            var incompleteSubSteps = SubSteps.Count(s => !s.IsDone);
+
+            return incompleteSubSteps == 0
+                ? new UpgradeStepInitializeResult(UpgradeStepStatus.Complete, "No Razor updaters need applied", BuildBreakRisk.None)
+                : new UpgradeStepInitializeResult(UpgradeStepStatus.Incomplete, $"{incompleteSubSteps} Razor updaters need applied", SubSteps.Where(s => !s.IsDone).Max(s => s.Risk));
+        }
+
+        public override UpgradeStepInitializeResult Reset()
+        {
+            _razorEngine = null;
+            _razorDocuments = null;
+            return base.Reset();
+        }
+
+        protected override Task<UpgradeStepApplyResult> ApplyImplAsync(IUpgradeContext context, CancellationToken token)
+        {
+            // Updates are made in sub-steps, so no changes need made in this apply step.
+            // Just check that sub-steps executed correctly.
+            var incompleteSubSteps = SubSteps.Count(s => !s.IsDone);
```

Comment:
Do we need to filter substeps by applicable?

Reply:
I don't think so. It probably wouldn't hurt, but inapplicable sub-steps should go straight to a completed state when they're initialized.